pkg/controllers/managementuser/rbac/roletemplates/roletemplate_handler.go

```diff
@@ -0,0 +1,258 @@
+// 2. an Aggregating ClusterRole that aggregates all inherited RoleTemplates' promoted Cluster Roles named "RoleTemplateName-promoted-aggregator"
+func (rtl *roleTemplateLifecycle) sync(rt *v3.RoleTemplate) (runtime.Object, error) {
+	// 1. Cluster role with rules
+	cr := rtl.buildClusterRole(clusterRoleNameFor(rt), rt.Name, rt.Rules)
+	err := rtl.createOrUpdateClusterRole(cr)
+	if err != nil {
+		return nil, err
+	}
+	// 2. Aggregating cluster role
+	cr = rtl.buildAggregatingClusterRole(clusterRoleNameFor(rt), rt.Name, rt.RoleTemplateNames)
+	err = rtl.createOrUpdateClusterRole(cr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Projects can have 2 extra cluster roles for global resources
+	if rt.Context == projectContext {
+		promotedRules := getPromotedRules(rt.Rules)
+
+		// If there are no promoted rules and no inherited RoleTemplates, no need for additional cluster roles
+		if len(promotedRules) == 0 && len(rt.RoleTemplateNames) == 0 {
+			return rt, nil
+		}
+
+		if len(promotedRules) != 0 {
+			// 3. Project global resources cluster role
+			cr = rtl.buildClusterRole(promotedClusterRoleNameFor(rt), rt.Name, promotedRules)
+			err = rtl.createOrUpdateClusterRole(cr)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// 4. Project global resources aggregating cluster role
+		// It's possible for this role to have no rules if there are no promoted rules in any of the inherited RoleTemplates or in the above ClusterRole (3)
+		// but without fetching all those RoleTemplates and looking through their rules, it's not possible to prevent this ahead of time as the Rules in
+		// an aggregating cluster role only get populated at run time
+		cr = rtl.buildAggregatingClusterRole(promotedClusterRoleNameFor(rt), rt.Name, rt.RoleTemplateNames)
+		err = rtl.createOrUpdateClusterRole(cr)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return rt, nil
+}
+
+// createOrUpdateClusterRole creates or updates the given Cluster Role
+func (rtl *roleTemplateLifecycle) createOrUpdateClusterRole(cr *v1.ClusterRole) error {
+	// attempt to get the cluster role
+	clusterRole, err := rtl.crClient.Get(cr.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// Cluster Role doesn't exist, create it
+		_, err = rtl.crClient.Create(cr)
+		if apierrors.IsAlreadyExists(err) {
+			// in the case where it got created at the same time somehow, attempt to get it again
+			clusterRole, err = rtl.crClient.Get(cr.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+		} else if err != nil {
+			return err
+		} else {
+			return nil
+		}
+	} else if err != nil {
+		return err
+	}
+
+	// check that the existing cluster role is the same as the one we want
+	if hasClusterRoleChanged(clusterRole, cr) {
+		// if it has changed, update it to the correct version
+		_, err := rtl.crClient.Update(cr)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
```

Comment:
In fact, I think we should create a generic `createOrUpdate` (e.g. accept client and object), as we may find more usages of this in your next PRs. Something like this maybe?:

```
type clientInterface[T runtime.Object] interface {
	Create(T) (T, error)
	GetNamespaced(namespace, name string, opts metav1.GetOptions) (T, error)
	Update(T) (T, error)
}

func createOrUpdate[T runtime.Object](client clientInterface[T], obj T, compareFunc func(T, T) bool) error {
	metadata, err := meta.Accessor(obj)
	if err != nil {
		return err
	}

	current, err := client.GetNamespaced(metadata.GetNamespace(), metadata.GetName(), metav1.GetOptions{})
	if err != nil {
		if !apierrors.IsNotFound(err) {
			return err
		}
		_, err := client.Create(obj)
		return err
	}

	if compareFunc(current, obj) {
		_, err := client.Update(obj)
		return err
	}
	return nil

}

// createOrUpdateClusterRole creates or updates the given Cluster Role
func (rtl *roleTemplateLifecycle) createOrUpdateClusterRole(cr *v1.ClusterRole) error {
	return createOrUpdate(rtl.crClient, cr, hasClusterRoleChanged)
}
```

Also, I'm not sure if using wrangler instead of norman clients would have something like that already.